Replace timestamp with navx_node

juliaschatz opened the pull request:
Instead of being a navX library with a basic executable, this will now be a ROS node that publishes IMU messages. There may be some additional work to do wrt ROS, but this at least compiles and runs.

juliaschatz committed:
Replace timestamp with navx_node

JWCS commented on package.xml:
```diff
@@ -0,0 +1,63 @@
+  <buildtool_depend>catkin</buildtool_depend>
+  <build_depend>sensor_msgs</build_depend>
+  <build_export_depend>sensor_msgs</build_export_depend>
+  <exec_depend>sensor_msgs</exec_depend>
```
These can be replaced with just an `<depend>sensor_msgs</depend>`

JWCS commented on package.xml:
```diff
@@ -0,0 +1,63 @@
+
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
```
Cleanup

JWCS commented on package.xml:
```diff
@@ -0,0 +1,63 @@
+  <!-- One license tag required, multiple allowed, one license per tag -->
+  <!-- Commonly used license strings: -->
+  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <license>TODO</license>
```
Cleanup; Also, most of the code here is from Kauai labs, which released it under the MIT license, which we also use. 
There is some ambiguity (well...) since some ppl from FRC900 added code without adding a license to theirs. But we can be safe enough just doing this.

JWCS commented on package.xml:
```diff
@@ -0,0 +1,63 @@
+
+  <!-- Author tags are optional, multiple are allowed, one per tag -->
+  <!-- Authors do not have to be maintainers, but could be -->
+  <!-- Example: -->
```
Delete

JWCS commented on package.xml:
```diff
@@ -0,0 +1,63 @@
+  <!-- Author tags are optional, multiple are allowed, one per tag -->
+  <!-- Authors do not have to be maintainers, but could be -->
+  <!-- Example: -->
+  <!-- <author email="jane.doe@example.com">Jane Doe</author> -->
```
The original authors of this as a package forked by us is Andrew & I, with our school emails. The maintainer(s) are the gofirst org. and, well, you if you want.

JWCS commented on package.xml:
```diff
@@ -0,0 +1,63 @@
+  <!-- Use test_depend for packages you need only for testing: -->
+  <!--   <test_depend>gtest</test_depend> -->
+  <!-- Use doc_depend for packages you need only for building documentation: -->
+  <!--   <doc_depend>doxygen</doc_depend> -->
```
I don't think we're doing any custom msgs, or anything fancy, but we will need roscpp

JWCS replied:
Also delete

JWCS commented on src/TimeStamp.cpp:
```diff
@@ -1,47 +0,0 @@
-#include <iostream>
```
Don't delete this file. This is a good example to demonstrate functionality without ros, and to test that the imu is working correctly (or what frame it's in, or do statistical analysis).

juliaschatz replied:
This still lives in the FRC900 repo as a standalone. It serves no purpose for us here.

JWCS replied:
It does. Trust me from experience, it is a nice way to verify that the hardware is working. Having a demo exe that can check/isolate issues with a hardware device is critical for saving time/energy. For example, we have cansend/dump exes for the canbus.

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+//    frequency (double): default=50.0
+//      The update frequency of the update loop
+//    param_name2 (param_name2_type): default=param_name2_default(,param_name1_path)
+//    param_name3 (param_name3_type): default=param_name3_default(,param_name1_path)
```
Delete

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+
+// ROS Topics
+std::string imu_topic = "imu/data";
+
```
Delete. It's slightly better form to just put the topics directly in the pub/sub in main.

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+// ROS Params
+double frequency = 50.0;
+//param_name2_type param_name2 = param_name2_default;
+//param_name3_type param_name3 = param_name3_default;
```
Delete

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+// int    global2 = 0;
+// std::string global3 = "";
+// void myCustomFunc1(int i);
+// bool myCustomFunc2(float f);
```
Delete

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+//param_name3_type param_name3 = param_name3_default;
+
+// Global_Vars
+AHRS com = AHRS("/dev/ttyACM0");
```
This should be a `AHRS* com;`
Initialization should be done in the main.

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+
+
+// ROS Node and Publishers
+ros::NodeHandle * nh;
```
For good style/less bugs, also add another NodeHandle:
`ros::NodeHandle * pnh;`
This "private" node handle we'll initialize to have local scope for params, and the normal node handle (nh) will have global scope (for topics/pubs/subs)

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+int main(int argc, char** argv){
+  // Init ROS
+  ros::init(argc, argv, "navx_node");
+  nh = new ros::NodeHandle("~");
```
Add/change:
`nh = new ros::NodeHandle(); pnh = new ros::NodeHandle("~");`

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+  ros::Timer update_timer = nh->createTimer(ros::Duration(1.0/frequency), update_callback);
+
+  // Publishers
+  imu_pub = nh->advertise<sensor_msgs::Imu>(imu_topic, 10);
```
Put the topic as specified in the docs above in here.

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+  // Params
+  nh->param<double>("frequency", frequency);
+  //nh->param<param_name2_type>(param_name2_path, param_name2, param_name2_default;
+  //nh->param<param_name3_type>(param_name3_path, param_name3, param_name3_default;
```
Delete

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+  imu_pub = nh->advertise<sensor_msgs::Imu>(imu_topic, 10);
+
+  // Params
+  nh->param<double>("frequency", frequency);
```
Ok, now for params, they should use `pnh` instead of `nh`.

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,106 @@
+  msg.linear_acceleration_covariance[6] = 1.569064;
+  
+  imu_pub.publish(msg);
+}
```
For com, need to change `com.method()` to `com->method()`

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
 
-include_directories(include)
+## Compile as C++11, supported in ROS Kinetic and newer
+# add_compile_options(-std=c++11)
```
See deleted line 4, where it had add_definitions(...)

JWCS reviewed (changes requested):
The tricky one here is the CMakeLists.txt
It needs to by default only do a normal compilation, unless compiled with catkin_make, to use it as both kinds of libraries.

juliaschatz commented:
> The tricky one here is the CMakeLists.txt
> It needs to by default only do a normal compilation, unless compiled with catkin_make, to use it as both kinds of libraries.

I'm not sure what you mean here. Do you foresee building this without Catkin? In what context? What needs to change for that to happen?

juliaschatz committed:
Minor cleanup and style changes

Remove extraneous comments. Change AHRS to pointer. Add separate node handle for params. Add compiler flags back in.

JWCS commented:
I originally saw your other comment first, which in light of I (kinda-mostly) agree with you wrt this file. 
The thing about the FRC900 repo, is that their code orginization is terrible, and not set up to be installed as a system library. When we originally grabbed it, before we used ros, we reformatted it such that we could install it (well). 
But now as we're using it chiefly as a ros node, I think the proper/simple way to handle it is to split of the current impl into a new branch called "no_ros_cmake", and make master fully embrace ros. 

In which case, you need to clean up the cmake file more.

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
-cmake_minimum_required (VERSION 2.8)
-project (NavX)
-set(CMAKE_BUILD_TYPE Release)
-add_definitions(-std=c++11 -Wall -Wextra -Wno-switch)
```
Want this

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
+#   cfg/DynReconf1.cfg
+#   cfg/DynReconf2.cfg
+# )
+
```
Delete this

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
+## DEPENDS: system dependencies of this project that dependent projects also need
+catkin_package(
+  INCLUDE_DIRS include
+#  LIBRARIES navx
```
Need to (ought to) expose our library

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
+catkin_package(
+  INCLUDE_DIRS include
+#  LIBRARIES navx
+  CATKIN_DEPENDS sensor_msgs
```
roscpp

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
+  INCLUDE_DIRS include
+#  LIBRARIES navx
+  CATKIN_DEPENDS sensor_msgs
+#  DEPENDS system_lib
```
Threads

juliaschatz replied:
Catkin seems to be angry about this because Threads doesn't define any standard include dirs: 
```
CMake Warning at /opt/ros/melodic/share/catkin/cmake/catkin_package.cmake:166 (message):
  catkin_package() DEPENDS on 'Threads' but neither 'Threads_INCLUDE_DIRS'
  nor 'Threads_LIBRARIES' is defined.
Call Stack (most recent call first):
  /opt/ros/melodic/share/catkin/cmake/catkin_package.cmake:102 (_catkin_package)
  navx/CMakeLists.txt:31 (catkin_package)
```

Is this safe to ignore?

JWCS replied:
Actually, instead of having multiple CATKIN_DEPENDS, you just have one with all the packages after it (doesn't have to be on same line, but it reads until the next separator (DEPENDS or end parenth) (this might've been the bug).
If you comment it out, that should be fine I think, as long as it links against it in the compiling. Easy test.

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
+# endif()
+
+## Add folders to be run by python nosetests
+# catkin_add_nosetests(test)
```
Delete

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
+## Add cmake target dependencies of the library
+## as an example, code may need to be generated before libraries
+## either from message generation or dynamic reconfigure
+# add_dependencies(${PROJECT_NAME} ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})
```
Need this

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,217 @@
+target_link_libraries(${PROJECT_NAME}_node
+   ${catkin_LIBRARIES}
+   navx
+ )
```
Clean up this

juliaschatz committed:
Clean up CMakeLists

juliaschatz committed:
Add license

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,85 @@
-#        PUBLIC_HEADER DESTINATION ${CMAKE_CURRENT_SOURCE_DIR}/build
-#)
+## Add folders to be run by python nosetests
+# catkin_add_nosetests(test)
```
Can be deleted

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,85 @@
+## same as for the library above
+add_dependencies(navx_node ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})
+
+## Specify libraries to link a library or executable target against
```
Delete the lines between add_executable/add_deps/target_link_libraries, and make them each one line. Wherever it says ${PROJECT_NAME} make it hardcoded navx.

JWCS commented on CMakeLists.txt:
```diff
@@ -1,60 +1,85 @@
+## INCLUDE_DIRS: uncomment this if your package contains header files
+## LIBRARIES: libraries you create in this project that dependent projects also need
+## CATKIN_DEPENDS: catkin_packages dependent projects also need
+## DEPENDS: system dependencies of this project that dependent projects also need
```
If you understand this, you can delete it. Same as the rest of the documentation.

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,97 @@
+
+int main(int argc, char** argv) {
+  // Init IMU
+  com = new AHRS("/dev/ttyACM0");
```
The proper order in main() is`{
// ROS Init
// Params
// Init (other stuff)
// Subscribers
// Publishers
// Spin
// Close/Delete any non-handled resources
}`
The order is because we need params usually for init'ing other stuff, and then subs before pubs because it matches the info flow: inputs to outputs. 
We don't have to clean up com after spin, b/c it's handled. But it should be after params, b/c params might set the device path/name, instead of hardcoding it (likewise for frequency - do you see how the timer is made with the default frequency instead of the parameter value?)

juliaschatz committed:
Continue cleaning up CMakeLists and re-order initialization in navx_node

Commit:
Merge branch 'master' into navx_node

juliaschatz committed:
Update README to be more accurate to this current version

juliaschatz committed:
Update attribution

JWCS reviewed (changes requested):
While I still think there is no point deleting the timestamp file, you should add a `device_path` param for the AHRS device path, a geometry_msgs point publisher of the euler angles (roll pitch yaw) published if `pub_euler_angles` param, on topic `imu/euler`, and params for defaults for the covariances for the variables should be made. 
Otherwise looks good.

juliaschatz commented:
Is there any point in not publishing the euler angles? Why would we want a param to turn that off/on?

juliaschatz commented:
Also, why publish Euler angles at all? They're not terribly useful since they carry the same data as the Quaternion data.

juliaschatz commented:
And reading the Imu message documentation, I don't think our Quaternion covariance should be -1- if we don't know, it should be 0.

juliaschatz commented:
Is there a way for us to estimate variance or covariance? If so, would we go through the effort of doing so?

JWCS commented on src/navx_node.cpp:
```diff
@@ -0,0 +1,97 @@
+
+void update_callback(const ros::TimerEvent&){
+  sensor_msgs::Imu msg;
+  
```
The msg needs a header, the frame_id needs to be a param. time can come from ros::time::now(), and seq can be a global variable incremented each loop.

JWCS commented:
1) It's good style to allow features to be turned on/off when not needed. 
2) The key point about euler angles is that they're human readable. From a computational standpoint, the quaternion is sufficient; yet for debugging and statistical testing of the actual standard deviations the euler angles are far more useful. 
3) The whole covar matrix can be noted as unknown/missing when element 0 is -1 (https://docs.ros.org/melodic/api/sensor_msgs/html/msg/Imu.html)
Also, you would probably be very interested in starting to learn about http://docs.ros.org/melodic/api/robot_localization/html/index.html
4) You're right - we ought to have an accurate value for at least the variance; either estimated or from data sheet / tuned. We will need it for the position estimation. I'm considering it out of scope for this PR because there's already heavy changes: numeric changes should be made in a separate commit/PR (as this PR will be squashed in merge).
5) I remembered something important: there should be a param for the frame_id, which led me to realize there wasn't a header for the imu msg. I made a comment about it. Just add `
msg.header.time = ros::Time::now();
msg.header.seq = seq++;
msg.header.frame_id = frame_id;` 
The param frame_id should by default be called "imu_link"

juliaschatz committed:
Add Euler message and Imu message header

juliaschatz committed:
Add pitch and roll rate measurements

juliaschatz commented:
```
# If the covariance of the measurement is known, it should be filled in (if all you know is the 
# variance of each measurement, e.g. from the datasheet, just put those along the diagonal)
# A covariance matrix of all zeros will be interpreted as "covariance unknown", and to use the
# data a covariance will have to be assumed or gotten from some other source
#
# If you have no estimate for one of the data elements (e.g. your IMU doesn't produce an orientation 
# estimate), please set element 0 of the associated covariance matrix to -1
# If you are interpreting this message, please check for a value of -1 in the first element of each 
# covariance matrix, and disregard the associated estimate.
```
I'm reading this as 
-1 in element 0: no orientation/velocity/etc measurement at all
0 in any element: unknown in covariance matrix or zero covariance?

In this case because we do provide all the measurements, we should have a full matrix of zeros (and then fill in variances along the diagonal when we measure them.)

Anything else I need to fix or is this good to go?

juliaschatz commented:
This probably should have a launch file, yeah?

JWCS commented:
Good job getting everything in the right format & style; and catching the covariance indices were off. 
You're right about the -1 in covariance invaluating the estimate, I had been misreading that as just invalidating the covars as whole.

JWCS commented:
The problem is, you deleted the most fundamentally critical part of the node, which is required for everything to work well 5 months from now: the documentation. 
The node _**NEEDS / MUST HAVE / REQUIRES**_ documentation at the top. 
This is critical when you start having 3-5 publishers/subscribers, depending on the formatting/data flow between components, and 50+ params to manage it all. 
When a package starts getting 5+ nodes, it similarly becomes impractical to throw all the docs between the file and the readme, and have to update default values in two files; while it's redundant info, it is superbly useful and peace-inducing. Since this package by nature only has one node, having the readme is a good idea, but in practice node documentation is a must. 
The true evil of ROS is the confusion and sparsity of documentation. Launch files are particularly terrifying; they're also all the real code.
Furthermore, there should be a launch file. It should have all the documentation as well. It can be kinda even more redundant, but the point is that someone needing to use the node can just copy/paste from the launch file, deleting params that don't need to be changed. There was an old one already (https://github.com/GOFIRST-Robotics/NASA-RMC-2019-NorthstarRobotics/blob/master/src/bigrdigr_bringup/launch/nodes/navx_node.launch), but the formatting should be more like
```
<?xml version="1.0" encoding="UTF-8"?> 
  <!-- 
  navx_node.launch
  VERSION: 0.0.0 
  LAST_CHANGED: 2019-04-01
  AUTHOR: 
  -->
<launch>
<!-- Interface: 
  Pub: imu/data (sensor_msgs/Imu)
  Param: frequency (double) 50.0; The frequency of the read loop
  -->
  <node pkg="bigrdigr_bringup" type="navx_node" name="navx_node" output="screen" respawn="true">
    <param name="frequency" value="50.0"/>
  </node>
</launch>
```
If it seems like I've gone on a bit about documentation, it's because I've done long term software dev with ros. It is the absolutely most important thing you can do.

juliaschatz commented:
Should the documentation include the subscription to a timer? This was listed in the original documentation but it looks totally internal to me.

juliaschatz commented:
I was able to run this tonight and the good news is that the node runs fine, except it doesn't publish Euler data. I think this is because I have the boolean set up incorrectly- I'll look at that soon.
It also doesn't actually publish any real navX data, which is weird but that's probably not a node problem.

JWCS commented:
Ah, yeah, about the timer, it doesn't really need to be there. 
If you make the changes I mentioned, I think it's good. If you push the last commit(s) I'll take a last look. 
The euler angles should help debug that? Do you have a navx with you? 
It really should pub the "real" data, otherwise we have a problem. A node should just be wrapping and encapsulating, boilerplate. It shouldn't mess up the data

juliaschatz committed:
Add launch file

juliaschatz committed:
Remove timer sub documentation

juliaschatz committed:
Change parameters so the default should work?

juliaschatz commented:
The problem is either in the navX comms code elsewhere or it was something wrong with the robot when I tested it. The node is fine. If you're okay with the launch file then let's go ahead and merge.

JWCS commented:
Two.5 last things:
0.5) Add a version to the launch file/node/package.xml if not already
1) This is probably nitpicky, but I disagree with your last commit, 613a3c1, moving the defaults. They should be closer to the top, easier to (inevitably) change with the documentation. If there seem to be wrong default/actual value shenanigans, thats because there's a subtle bug/feature somewhere with how the params are done. I'm pretty positive the way you had it won't have any problems (as that's how all of my other stuff has worked without issue). 
2) Test it out by running it on your computer, with the navx plugged in directly, not indirectly through the robot. Make sure there are no problems, and then I'm cool. Software without testing is time lost later on.

JWCS commented:
Also, in the docs in the launch file, put the pubs/params on new line below Pub: and Param:, with only one level of indentation, not however much for "Pub:"/"Param:"

juliaschatz committed:
Update indentation in launch file

juliaschatz committed:
Set version to 1.0.0

juliaschatz commented:
Tested it just now on my computer- pubs IMU data and euler angles if we turn that on. Will have to figure out why it's not working on the Jetson.

Personally I find this style of default params easier to read. I think any more arguing about this is easily into bikeshedding, since both work and it's almost entirely preference.

Added version 1.0.0, since this is ideally fully featured and working properly.

juliaschatz commented:
Jetson issue isn't this package, so we're good to merge.

JWCS commented:
I'll disagree, but merge-squash it. (If the changes inbetween were more important, perhaps not, or rebased if someone else did work too)
